sourcegraph/sourcegraph
# sveltekit: Add history panel

https://github.com/sourcegraph/sourcegraph/assets/179026/373b3158-eb62-467c-8198-eda0ac654791




This PR adds a bottom panel to the repo root, file and directory pages that when opened shows the respective commit history. The commits are lazily loaded, as the list is scrolled more data is loaded. With SvelteKit's snapshot feature the loading state (i.e. which commits have been loaded), the scroll position as well as the panel's open/close state are associated with the browser history. So as one is going through the browser history with the back and forth button we are restoring the exact same state that the page was in.

There are two important things to point out: 

1. Whether storing information about the loaded commits as a snapshot is a good idea remains to be seen. SvelteKit stores that data in session storage which has limited space.
2. Restoring only works for the browser history. If navigate to a previously opened page in a different way (e.g. by clicking the file tree entry) then it would load in its initial state. If we also wanted to restore the loaded commits and scroll position we would have to implement our own logic for this.

Other things that are going on in this PR:

- The panel is not resizable yet. The `Seperator` component should be rewritten in a separate PR to add support for vertical resizing.
- Added a `Scroller` component to make it easier to manage scroll state and lazy loading. I got this from a more complex example that also included virtualizing the list (i.e. only rendering a subset of items). I hope to add that functionality to `Scroller` as well some day.
- The existing `Tabs` component was fixed to properly remove existing Tabs when the component (or its parent) gets hot reloaded. It was also extended to allow for a "no tab selected" state and to control the selected tab from the parent.
- `UserAvatar` was extended to render avatar URLs. I added stories for the different states.
- The blob and tree pages have been extended to show diff information instead of the file content/list when the URL contains a `rev` parameter. It was an easy way to implement it, but it also means that every time we select a revision and update the URL we also (in theory) refetch the blob and tree data. At the moment we are memoizing this data, so no actual network calls happen, but we might find a better approach in the future (e.g. when shallow routing becomes available)
- I had some type problems and so I changed again how `load` functions return "lazy loaded" data (`{deferred: {data: ...}}` instead of `{data: {deferred: ...}}`). This adds a bit of noise here.
- I added `testdata.ts` which uses faker to generate more realistic example data for GraphQL-related objects. It's used for the history panel stories. Eventually I'd like to have such example builders for every GraphQL object used in the app, for unit tests and stories (ideally the examples could be automatically generated from the schema)

## Test plan

- Open history panel on repo root, file and directory page.
  - Select a commit and verify that the diff view is shown on repo and directory pages.
  - Use the browser history buttons to verify that scroll state is restored.
- Open other repo pages to make sure that they still work.
